Submission client redesign to use a step-based builder pattern

mccheah opened the pull request:
Applies changes assuming PySpark is present from #364 .

This change overhauls the underlying architecture of the submission client, but it is intended to entirely preserve existing behavior of Spark applications. Therefore users will find this to be an invisible change.

The philosophy behind this design is to reconsider the breakdown of the submission process. It operates off the abstraction of "submission steps", which are transformation functions that take the previous state of the driver and return the new state of the driver. The driver's state includes its Spark configurations and the Kubernetes resources that will be used to deploy it.

Such a refactor moves away from a features-first API design, which considers different containers to serve a set of features. The previous design, for example, had a container files resolver API object that returned different resolutions of the dependencies added by the user. However, it was up to the main Client to know how to intelligently invoke all of those APIs. Therefore the API surface area of the file resolver became untenably large and it was not intuitive of how it was to be used or extended.

This design changes the encapsulation layout; every module is now responsible for changing the driver specification directly. An orchestrator builds the correct chain of steps and hands it to the client, which then calls it verbatim. The main client then makes any final modifications that put the different pieces of the driver together, particularly to attach the driver container itself to the pod and to apply the Spark configuration as command-line arguments.

The current steps are:

1. `aseSubmissionStep`: Baseline configurations such as the docker image and resource requests.
2. `DriverKubernetesCredentialsStep`: Resolves Kubernetes credentials configuration in the driver pod. Mounts a secret if necessary.
3. `InitContainerBootstrapStep`: Attaches the init-container, if necessary, to the driver pod. This is optional and wont' be loaded if all URIs are "local" or there are no URIs at all.
4. `DependencyResolutionStep`: Sets the classpath, spark.jars, and spark.files properties. This step is partially not isolated as it assumes that files that are remote or locally submitted will be downloaded to a given location. Unit tests should verify that this contract holds.
5. `PythonStep`: Configures Python environment variables if using PySpark.

mccheah commented:
Rebuilt from #363.

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/PythonStep.scala:
```diff
@@ -27,15 +27,19 @@ private[spark] class PythonStep(
     filesDownloadPath: String) extends KubernetesSubmissionStep {
 
   override def prepareSubmission(driverSpec: KubernetesDriverSpec): KubernetesDriverSpec = {
+    val resolvedOtherPyFilesString = if (otherPyFiles.isEmpty) {
```
@ifilonenko is this sufficient to cover the arguments handling? SPARK_DRIVER_ARGS is loaded in `BaseSubmissionStep` already. Or is there more we need to do here?

ifilonenko replied:
Not exactly, it would need to be `null` unless you have docker parse, but docker doesnt do if: else: blocks so thats why this woudn't be helpful per say

mccheah replied:
Ok - but if we changed this to be the string "null" then that would suffice?

mccheah replied:
We can have the Dockerfile use an if-else block also - we do this in a number of places in the existing ones. It's just bash if-else syntax in the command.

ifilonenko replied:
I am testing that rn in a seperate PR, but yes, it should. passing in just `null` I believed cause withValue() to raise an error. But I think that CMD PythonRunner PY_FILE "null" DRIVER_ARGS parses correctly.

ifilonenko replied:
if-else syntax doesnt exist? there is only if - then

mccheah replied:
Bash supports if-then-else. It might be tricky to get it exactly right in the Dockerfile so maybe it's not worth it.

ifilonenko commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -621,17 +621,12 @@ object SparkSubmit {
-        childArgs += "org.apache.spark.deploy.PythonRunner"
-        childArgs += "--other-py-files"
-        childArgs += args.pyFiles
+        childArgs ++= Array("--primary-py-file", args.primaryResource)
```
:) thanks for the --primary-py-file

ifilonenko commented on resource-managers/kubernetes/integration-tests/src/test/scala/org/apache/spark/deploy/kubernetes/integrationtest/backend/minikube/MinikubeTestBackend.scala:
```diff
@@ -27,7 +27,7 @@ private[spark] class MinikubeTestBackend extends IntegrationTestBackend {
   override def initialize(): Unit = {
     Minikube.startMinikube()
-    new SparkDockerImageBuilder(Minikube.getDockerEnv).buildSparkDockerImages()
+//    new SparkDockerImageBuilder(Minikube.getDockerEnv).buildSparkDockerImages()
```
Intentionally commited?

mccheah replied:
Nope good catch - was from local testing

mccheah commented:
Conflicts are likely from the style changes in the base PR. I expect that unless we have significant deviations functionality-wise in the Python implementation, we can resolve most if not all of the conflicts by just taking this branch.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,85 @@
+  @Mock
+  private var podWithDetachedInitContainer : PodWithDetachedInitContainer = _
+  @Mock
+  private var initContainerSpec : InitContainerSpec = _
```
Don't use a mock here - since it's a case class, the equivalent of a struct, it's fine to just use the implementation.

ifilonenko replied:
How would you simulate the .copy method tho?

mccheah replied:
We don't need to - since it's a Scala primitive we should operate under what its actual behavior is. We don't mock classes like `scala.collection.List` or `java.util.Optional` for similar reasons.

ifilonenko replied:
Agreed. I will just pass in an empty case class

mccheah commented:
I resolved merge conflicts with the "ours" strategy since the only difference in the latest push that caused the conflicts were style changes. I'll rebase this branch to also be pointing to `branch-2.1-kubernetes` shortly.

mccheah committed:
Submission client redesign to use a step-based builder pattern.

This change overhauls the underlying architecture of the submission
client, but it is intended to entirely preserve existing behavior of
Spark applications. Therefore users will find this to be an invisible
change.

The philosophy behind this design is to reconsider the breakdown of the
submission process. It operates off the abstraction of "submission
steps", which are transformation functions that take the previous state
of the driver and return the new state of the driver. The driver's state
includes its Spark configurations and the Kubernetes resources that will
be used to deploy it.

Such a refactor moves away from a features-first API design, which
considers different containers to serve a set of features. The previous
design, for example, had a container files resolver API object that
returned different resolutions of the dependencies added by the user.
However, it was up to the main Client to know how to intelligently
invoke all of those APIs. Therefore the API surface area of the file
resolver became untenably large and it was not intuitive of how it was
to be used or extended.

This design changes the encapsulation layout; every module is now
responsible for changing the driver specification directly. An
orchestrator builds the correct chain of steps and hands it to the
client, which then calls it verbatim. The main client then makes any
final modifications that put the different pieces of the driver
together, particularly to attach the driver container itself to the pod
and to apply the Spark configuration as command-line arguments.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/InitContainerStepsOrchestratorSuite.scala:
```diff
@@ -0,0 +1,82 @@
+  private val STAGING_SERVER_URI = "http://localhost:8000"
+
+  test ("Contact resource staging server w/o TLS") {
+    val SPARK_CONF = new SparkConf(true)
```
Call this `sparkConf`

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/InitContainerStepsOrchestratorSuite.scala:
```diff
@@ -0,0 +1,82 @@
+
+    val initSteps : Seq[InitContainerStep] = initContainerStepsOrchestrator.getInitContainerSteps()
+    assert(initSteps.length == 2)
+    assert( initSteps.map({
```
This is just more cleanly expressed as

```
assert(initSteps(0).isInstanceOf[BaseInitContainerStep])
assert(initSteps(1).isInstanceOf(SubmittedResourcesInitContainerStep])
```

instead of using pattern matching.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/InitContainerStepsOrchestratorSuite.scala:
```diff
@@ -0,0 +1,82 @@
+      INIT_CONTAINER_CONFIG_MAP_NAME, INIT_CONTAINER_CONFIG_MAP_KEY, SPARK_CONF)
+    val initSteps : Seq[InitContainerStep] = initContainerStepsOrchestrator.getInitContainerSteps()
+    assert(initSteps.length == 1)
+    assert(initSteps.headOption.exists({
```
Similarly here, just use `isInstanceOf`.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,78 @@
+  @Mock
+  private var podAndInitContainerBootstrap : SparkPodInitContainerBootstrap = _
+  @Mock
+  private var podWithDetachedInitContainer : PodWithDetachedInitContainer = _
```
This is a case class so it shouldn't be a mock.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,78 @@
+      podAndInitContainerBootstrap)
+    val remoteJarsToDownload = KubernetesFileUtils.getOnlyRemoteFiles(SPARK_JARS)
+    val remoteFilesToDownload = KubernetesFileUtils.getOnlyRemoteFiles(SPARK_FILES)
+    assert(remoteJarsToDownload === List("hdfs://localhost:9000/app/jars/jar1.jar"))
```
Why this check? We aren't testing the functionality of `KubernetesFileUtils` here.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,78 @@
+      new Container(), new Container(), new Pod, Seq.empty[HasMetadata]
+    )
+    val returnContainerSpec = baseInitStep.prepareInitContainer(initContainerSpec)
+    assert(expectedTest.toSet.subsetOf(returnContainerSpec.initContainerProperties.toSet))
```
Check for an exact match. We don't want to be setting unexpected properties here.

mccheah replied:
We also should be inspecting the properties of the pod and the containers. The mock of the `SparkPodInitContainerBootstrap` instance we pass in should modify the Kubernetes components in such a way that we can check for them afterwards. Basically we want to verify that the pod init container bootstrap was used to make changes to the pod and containers.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,78 @@
+    assert(remoteJarsToDownload === List("hdfs://localhost:9000/app/jars/jar1.jar"))
+    assert(remoteFilesToDownload === List("hdfs://localhost:9000/app/files/file1.txt"))
+    val expectedTest = Map(
+      INIT_CONTAINER_JARS_DOWNLOAD_LOCATION.key -> JARS_DOWNLOAD_PATH,
```
Could be a better name for this - `expectedDriverSparkConf`, perhaps.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/InitContainerStepsOrchestratorSuite.scala:
```diff
@@ -0,0 +1,82 @@
+      .set(RESOURCE_STAGING_SERVER_URI, STAGING_SERVER_URI)
+
+    val initContainerStepsOrchestrator = new InitContainerStepsOrchestrator(
+      NAMESPACE, APP_RESOURCE_PREFIX, SPARK_JARS, SPARK_FILES, JARS_DOWNLOAD_PATH,
```
One argument per line.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/InitContainerStepsOrchestratorSuite.scala:
```diff
@@ -0,0 +1,82 @@
+  private val INIT_CONTAINER_CONFIG_MAP_KEY = "spark-init-config-map-key"
+  private val STAGING_SERVER_URI = "http://localhost:8000"
+
+  test ("Contact resource staging server w/o TLS") {
```
Probably no need to mention TLS in the test description since it's largely irrelevant for what the test is actually checking.

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -16,256 +16,130 @@
+      driverContainer = new ContainerBuilder().build(),
+      driverSparkConf = submissionSparkConf.clone(),
+      otherKubernetesResources = Seq.empty[HasMetadata])
+    // This orchestrator determines which steps are necessary to take to resolve varying
```
This isn't really the orchestrator - the orchestrator has pre-determined these steps to run. Perhaps this comment can be moved?

mccheah committed:
Add a unit test for BaseSubmissionStep.

ifilonenko commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,8 @@
+package org.apache.spark.deploy.kubernetes.submit.submitsteps.initcontainer
```
Total accident here that I included this. derp

mccheah committed:
Add unit test for kubernetes credentials mounting.

erikerlandson commented on core/src/main/scala/org/apache/spark/deploy/SparkSubmit.scala:
```diff
@@ -621,14 +621,17 @@ object SparkSubmit {
-        childArgs += args.primaryResource
-        childArgs += "org.apache.spark.deploy.PythonRunner"
-        childArgs += args.pyFiles
+        childArgs ++= Array("--primary-py-file", args.primaryResource)
```
Wondering if it makes sense for `(childArgs, childClasspath, sysProps, childMainClass)` to be modeled as a case class with builder pattern, for similar reasons.  Tangentially, mutable collections can be hazardous if not handled carefully - a case-class pattern using immutable collections might be worthwhile, given the complexity of the environment constructions

mccheah replied:
We don't have many better options here because `SparkSubmit` creates the submission client implementation reflectively and only expects the submission client to have a main method with a list of arguments. This is to account for the fact that the core module of Spark doesn't have a compile time dependency on the specific submission client implementations for the different cluster managers.

mccheah replied:
In `Client.scala` we parse the arguments array into a case class and report on errors when fields are missing.

mccheah committed:
Add unit test for InitContainerBootstrapStep.

ifilonenko committed:
unit tests for initContainer

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -37,16 +36,28 @@ class BaseInitContainerStepSuite extends SparkFunSuite with BeforeAndAfter{
-      any[PodWithDetachedInitContainer])).thenReturn(podWithDetachedInitContainer)
+      any[PodWithDetachedInitContainer])).thenReturn(
+      PodWithDetachedInitContainer(
+        new PodBuilder()
```
We probably want to use the "base" spec here too. See for example the test implementations of the submission steps in `ClientSuite`.

ifilonenko replied:
Why not just check a label to see that it is indeed the one updated in the method?

mccheah replied:
The way the test is written right now makes it such that if the step made modifications to the pod before passing the pod along to the bootstrap plugin, those changes would be lost in the test code.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -73,6 +80,9 @@ class BaseInitContainerStepSuite extends SparkFunSuite with BeforeAndAfter{
     )
     val returnContainerSpec = baseInitStep.prepareInitContainer(initContainerSpec)
-    assert(expectedTest.toSet.subsetOf(returnContainerSpec.initContainerProperties.toSet))
+    assert(expectedDriverSparkConf.toSet === returnContainerSpec.initContainerProperties.toSet)
```
Nit: Can compare the maps directly without converting with `toSet`.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,123 @@
+  }
+  private val STAGING_SERVER_URI = "http://localhost:8000"
+  private val SECRET_MOUNT_PATH = "/tmp"
+  private val RSS_SECRET = createRSecret
```
Inline this. Could be

```
private val RSS_SECRET = Map(INIT_CONTAINER_SUBMITTED_JARS_SECRET_KEY -> BaseEncoding...)
```

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,123 @@
+      SubmittedResourceIdAndSecret(FILES_RESOURCE_ID, FILES_SECRET)
+    )
+    when(submittedResourcesSecretPlugin.mountResourceStagingServerSecretIntoInitContainer(
+      mockitoEq(new Container()))).thenReturn(
```
If `Container` doesn't properly implement `.equals()` properly then this will not match properly. We can probably just use `any()` here rather reasonably.

mccheah replied:
Or match on the specific instance of the `Container` that will be passed to this component.

ifilonenko replied:
I thought to do that, but I thought that an empty instance would suffice. But I guess `any(Container)` is reasonable

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,123 @@
+      mockitoEq(new Container()))).thenReturn(
+        new ContainerBuilder().withName("mountedSecret").build())
+    when(submittedResourcesSecretPlugin.addResourceStagingServerSecretVolumeToPod(
+      mockitoEq(new Pod()))).thenReturn(
```
Similarly here - can match on `any()` or else try to capture the actual reference to the Pod object that would be passed in here.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,123 @@
+      STAGING_SERVER_URI,
+      SECRET_MOUNT_PATH,
+      false,
+      None,
```
Probably worth setting these to `Some(...)` so that we inspect that they're set properly in the resolved Spark configuration. Or, write a separate test for specifically checking how we mount the secrets for some of these fields.

mccheah replied:
This is done below.

mccheah committed:
Add a unit test for DependencyResolutionStep.

ifilonenko committed:
further modifications to InitContainer unit tests

ifilonenko committed:
Merge branch 'submission-steps-refactor' of https://github.com/apache-spark-on-k8s/spark into submission-steps-refactor

ifilonenko committed:
Use of resolver in PythonStep and unit tests for PythonStep

mccheah commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/PythonStep.scala:
```diff
@@ -30,7 +30,7 @@ private[spark] class PythonStep(
       "null"
     } else {
-      otherPyFiles.mkString(",")
+      KubernetesFileUtils.resolveSubmittedUris(otherPyFiles, filesDownloadPath).mkString(",")
```
Does `PythonRunner` handle files with URI schemes of `hdfs` or other remote locations? If not, then we probably want to use the downloaded path for every URI - hence we should use `KubernetesFileUtils.resolveFilePaths` instead.

ifilonenko replied:
I believe it should. Why shouldn't it?

mccheah replied:
I'm not aware if the PythonRunner handles the downloading of remote files. However we also don't want to be downloading twice and if the init-container is already downloading the Python files, it might be worthwhile to only refer to the pre-downloaded files anyways.

ifilonenko replied:
I understand what you are saying. Although, when I did testing before I didn't see the remote files downloaded into `/var/spark-data/`. Is this handled with this PR? If it is, then, yeah it should used resolveFilePaths()

mccheah replied:
Hm, the remote files should most definitely be downloaded there. This was the case before this change.

mccheah replied:
Remote jars had to have been downloaded for the KubernetesSuite integration tests to succeed. I would expect remote files to be downloaded identically.

ifilonenko replied:
I saw that to be the case with `file://` not with `hdfs://` but I will investigate. You are write about resolveFilePaths() tho

ifilonenko commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,173 @@
+    Map(
+      RESOURCE_STAGING_SERVER_TRUSTSTORE_PASSWORD.key -> TRUSTSTORE_PASS,
+      RESOURCE_STAGING_SERVER_TRUSTSTORE_TYPE.key -> TRUSTSTORE_TYPE,
+      RESOURCE_STAGING_SERVER_TRUSTSTORE_FILE.key -> (SECRET_MOUNT_PATH + "/trustStore"),
```
Why does 
RESOURCE_STAGING_SERVER_TRUSTSTORE_FILE resort to 
`SECRET_MOUNT_PATH + "/trustStore"`
and 
RESOURCE_STAGING_SERVER_CLIENT_CERT_PEM resort to 
`SECRET_MOUNT_PATH + "/cert"`
by default here? Is that intentional?

ifilonenko replied:
I believe it stems from 
`INIT_CONTAINER_STAGING_SERVER_TRUSTSTORE_SECRET_KEY`
`INIT_CONTAINER_STAGING_SERVER_CLIENT_CERT_SECRET_KEY`
Reason for this?

mccheah replied:
The logic here mounts the user's secret files as a secret volume into the pod. When you create a secret, you provide a map of key-value pairs. One of these keys for the trustStore is `trustStore` (see `constants.scala`). When secrets are mounted onto the disk, one provides a base path and a file is created for every secret data item in the secret, with the file name being the secret's key. Since we mount the secret at `SECRET_MOUNT_PATH` and the secret key is `trustStore`, the resolved path is `SECRET_MOUNT_PATH/trustStore`.

To make this clearer, use the constant from `constants.scala` in the unit test.

mccheah replied:
Also could be worth putting a comment in the code to clarify this for future readers.

ifilonenko replied:
I will update it with the constants, I only just now realized after looking the code again. Yeah a comment would be key

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,173 @@
+    INIT_CONTAINER_SUBMITTED_FILES_SECRET_KEY ->
+      BaseEncoding.base64().encode(FILES_SECRET.getBytes(Charsets.UTF_8))
+  ).asJava
+  private val TRUSTSTORE_FILE = createTempFile(".jks")
```
I think it's idiomatic to create these files in a `before` block and to delete them with an `after` block. If one does this, the specific name of the file doesn't have to be a `UUID`.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,173 @@
+        EXECUTOR_INIT_CONTAINER_SECRET_MOUNT_DIR.key -> SECRET_MOUNT_PATH))
+  }
+
+  test ("testing prepareInitContainer w/ CERT and TrustStore Files w/o SSL") {
```
This test checks for when we're sending files from the submitter's local disk to mount as a secret. But the user is also allowed to specify these trustStore paths as `local://` paths if they're injecting their certificates into the init-container in a different way. We should have a unit test for that.

mccheah replied:
Looks like that's done below.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/InitContainerStepsOrchestratorSuite.scala:
```diff
@@ -0,0 +1,92 @@
+      INIT_CONTAINER_CONFIG_MAP_KEY,
+      sparkConf)
+    val initSteps : Seq[InitContainerStep] = initContainerStepsOrchestrator.getInitContainerSteps()
+    assert(initSteps.length == 1)
```
Nit: `===` instead of `==`.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,88 @@
+      INIT_CONTAINER_REMOTE_FILES.key -> "hdfs://localhost:9000/app/files/file1.txt"
+    )
+    val initContainerSpec = InitContainerSpec(
+      Map.empty[String, String], Map.empty[String, String],
```
Nit: Every argument on its own line.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,88 @@
+    MockitoAnnotations.initMocks(this)
+    when(podAndInitContainerBootstrap.bootstrapInitContainerAndVolumes(
+      any[PodWithDetachedInitContainer])).thenReturn(
+      PodWithDetachedInitContainer(
```
Use an `Answer` and wrap the PodBuilder that's passed in.

ifilonenko replied:
Can I just take the Pod that is passed in and run a .copy() with my own values?

ifilonenko committed:
refactoring of init unit tests and pythonstep resolver logic

mccheah committed:
Add unit test for KubernetesSubmissionStepsOrchestrator.

mccheah commented:
rerun unit tests please

mccheah commented:
rerun unit tests please

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/InitContainerStepsOrchestratorSuite.scala:
```diff
@@ -0,0 +1,92 @@
+  test ("including step to contact resource staging server") {
+    val sparkConf = new SparkConf(true)
+      .set(KUBERNETES_DRIVER_LABELS, s"$DEPRECATED_CUSTOM_LABEL_KEY=$DEPRECATED_CUSTOM_LABEL_VALUE")
+      .set(s"$KUBERNETES_DRIVER_LABEL_PREFIX$CUSTOM_LABEL_KEY", CUSTOM_LABEL_VALUE)
```
I think we need a `.` between the label prefix and the label key for this to parse properly.

mccheah replied:
Never mind, the dot is in the prefix in the constants.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,228 @@
+    )
+    when(submittedResourcesSecretPlugin.mountResourceStagingServerSecretIntoInitContainer(
+      any[Container])).thenReturn(
+        new ContainerBuilder().withName("mountedSecret").build())
```
Make sure we're preserving the contents of the container that was passed in.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,228 @@
+        new ContainerBuilder().withName("mountedSecret").build())
+    when(submittedResourcesSecretPlugin.addResourceStagingServerSecretVolumeToPod(
+      any[Pod])).thenReturn(
+      new PodBuilder()
```
Similarly - preserve the pod that was passed in.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,228 @@
+      INIT_CONTAINER_DOWNLOAD_FILES_RESOURCE_IDENTIFIER.key -> FILES_RESOURCE_ID,
+      INIT_CONTAINER_DOWNLOAD_FILES_SECRET_LOCATION.key ->
+        s"$SECRET_MOUNT_PATH/$INIT_CONTAINER_SUBMITTED_FILES_SECRET_KEY",
+      RESOURCE_STAGING_SERVER_SSL_ENABLED.key -> false.toString) ++
```
No need to combine two maps.

ifilonenko replied:
also will handle with next push

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,228 @@
+        RESOURCE_STAGING_SERVER_CLIENT_CERT_PEM.key ->
+          "/tmp/cert.pem"
+      )
+    assert(returnedInitContainer.initContainerProperties === expectedinitContainerProperties)
```
We probably want to inspect the other parts of the InitContainerSpec that are returned.

ifilonenko replied:
That spec would be identical to other tests above, no need, changing to local:// only affects the initContainerProperties

mccheah replied:
If the trustStore URI is `local://` then it shouldn't be in the secret object.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/BaseInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,101 @@
+          val pod = invocation.getArgumentAt(0, classOf[PodWithDetachedInitContainer])
+          pod.copy(
+            pod =
+              new PodBuilder()
```
Don't lose the original pod here. Use

`pod = new PodBuilder(pod.pod)...`

ifilonenko replied:
good catch

ifilonenko replied:
Will handle with next push, forgot to add

mccheah commented:
Still have a few unit tests to finish but the idea is there now. Sorry the review is so large - I thought about how it could be broken down into smaller parts, but it's hard to have the old implementation co-exist with a partial new one.

ifilonenko committed:
refactoring and addition of secret trustStore+Cert checks in a SubmissionStepSuite

ifilonenko committed:
added SparkPodInitContainerBootstrapSuite

ifilonenko committed:
Added InitContainerResourceStagingServerSecretPluginSuite

mccheah commented:
rerun integration tests please

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/SparkPodInitContainerBootstrapSuite.scala:
```diff
@@ -16,151 +16,73 @@
+    INIT_CONTAINER_CONFIG_MAP_NAME,
+    INIT_CONTAINER_CONFIG_MAP_KEY
+  )
+  private val returnedPodWithCont = sparkPodInit.bootstrapInitContainerAndVolumes(
```
Move these into the test method.

mccheah replied:
Sharing fields is nice for the object under test, but for the functionality we specifically are testing, it's more idiomatic to put the method calls into the test.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/SparkPodInitContainerBootstrapSuite.scala:
```diff
@@ -16,151 +16,73 @@
+      initContainer = new Container(),
+      mainContainer = new ContainerBuilder().withName(MAIN_CONTAINER_NAME).build())
+  )
+  private val expectedSharedMap = Map(
```
Clearer name, related to volumes

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/InitContainerResourceStagingServerSecretPluginSuite.scala:
```diff
@@ -0,0 +1,61 @@
+  private val INIT_CONTAINER_SECRET_NAME = "init-secret"
+  private val INIT_CONTAINER_SECRET_MOUNT = "/tmp/secret"
+
+  private val initContainerRSSP = new InitContainerResourceStagingServerSecretPluginImpl(
```
`initContainerSecretPlugin` - no need for an acronym.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/InitContainerResourceStagingServerSecretPluginSuite.scala:
```diff
@@ -0,0 +1,61 @@
+  test("Volume Mount into InitContainer") {
+    val returnedCont = initContainerRSSP.mountResourceStagingServerSecretIntoInitContainer(
+      new ContainerBuilder().withName("init-container").build()
+    )
```
This closing bracket should be on the previous line. There's a few places where this is done, so please fix the others also.

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/InitContainerResourceStagingServerSecretPluginSuite.scala:
```diff
@@ -0,0 +1,61 @@
+    assert(returnedCont.getVolumeMounts.asScala.map(
+      vm => (vm.getName, vm.getMountPath)) ===
+        List((INIT_CONTAINER_SECRET_VOLUME_NAME, INIT_CONTAINER_SECRET_MOUNT))
+    )
```
Bracket on the previous line

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/SparkPodInitContainerBootstrapSuite.scala:
```diff
@@ -16,151 +16,73 @@
+    DOWNLOAD_TIMEOUT_MINUTES,
+    INIT_CONTAINER_CONFIG_MAP_NAME,
+    INIT_CONTAINER_CONFIG_MAP_KEY
+  )
```
Bracket on the previous line

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/SparkPodInitContainerBootstrapSuite.scala:
```diff
@@ -16,151 +16,73 @@
+  private val expectedSharedMap = Map(
+    JARS_DOWNLOAD_PATH -> INIT_CONTAINER_DOWNLOAD_JARS_VOLUME_NAME,
+    FILES_DOWNLOAD_PATH -> INIT_CONTAINER_DOWNLOAD_FILES_VOLUME_NAME
+  )
```
Bracket on the previous line

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,251 @@
+import org.scalatest.BeforeAndAfter
+
+class SubmittedResourcesInitContainerStepSuite extends SparkFunSuite with BeforeAndAfter {
+  private def createTempFile(extension: String): String = {
```
Move this method to the bottom of the test class

mccheah commented on resource-managers/kubernetes/core/src/test/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/initcontainer/SubmittedResourcesInitContainerStepSuite.scala:
```diff
@@ -0,0 +1,251 @@
+          .addToLabels("mountedSecret", "true")
+          .endMetadata()
+          .withNewSpec().endSpec()
+          .build()}})
```
Closing curly braces should be on their own lines.

ifilonenko committed:
style in Unit tests

ifilonenko committed:
extremely minor style fix in variable naming

ifilonenko reviewed (approved):
LGTM

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/SparkPodInitContainerBootstrap.scala:
```diff
@@ -41,13 +47,12 @@ private[spark] class SparkPodInitContainerBootstrapImpl(
-      mainContainerName: String,
-      originalPodSpec: PodBuilder): PodBuilder = {
+      originalPodWithUnattachedInitContainer: PodWithDetachedInitContainer)
+      : PodWithDetachedInitContainer = {
```
this syntax looks a bit weird -- should the `:` be on the prior line?

mccheah replied:
It doesn't fit on one line in under 100 characters.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/config.scala:
```diff
@@ -418,6 +418,14 @@ package object config extends Logging {
       .createOptional
 
+  private[spark] val INIT_CONTAINER_REMOTE_PYSPARK_FILES =
+    ConfigBuilder("spark.kubernetes.initcontainer.remotePyFiles")
```
why is this newly-required from this refactor?  I expected there to be change in user-visible behavior

mccheah replied:
Good catch - I don't think this is necessary, this is an artifact of something I was trying before.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -16,256 +16,130 @@
+    var mainClass: Option[String] = None
+    val driverArgs = mutable.Buffer.empty[String]
+    args.sliding(2).toList.collect {
+      case Array("--primary-py-file", mainPyFile: String) =>
```
is this a new flag?

mccheah replied:
Yep - the contract for arguments sent to the child submission client class have changed.

ash211 replied:
why?  I thought we wanted no user-visible changes?

ash211 replied:
Seems like a lot of these are to do the `.sliding(2)` thing

mccheah replied:
It's not user-visible because this class is proxied into from `SparkSubmit.scala`.

mccheah replied:
We shouldn't expect this class to be used directly.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -16,256 +16,130 @@
+        throw new RuntimeException(s"Unknown arguments: $other")
+    }
+    require(mainAppResource.isDefined,
+        "Main app resource must be defined by either --py-file or --main-java-resource.")
```
do you mean `--primary-py-file` and `--primary-java-resource` here?

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -16,256 +16,130 @@
+   /**
+    * Run command that initalizes a DriverSpec that will be updated
+    * after each KubernetesSubmissionStep in the sequence that is passed in.
+    * The final driver-spec will be used to build the Driver Container,
```
`s/driver-spec/DriverSpec/` -- this is the only place it's hyphenated

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -16,256 +16,130 @@
+    var currentDriverSpec = KubernetesDriverSpec(
+      driverPod = basePod,
+      driverContainer = new ContainerBuilder().build(),
+      driverSparkConf = submissionSparkConf.clone(),
```
seems like only `submissionSparkConf` is different on each run -- create a constructor object for the `KubernetesDriverSpec` case class that takes just a SparkConf and contains the rest of this boilerplate.  should simplify tests a bit too

mccheah replied:
I think we want to enforce that all submission steps would specify all of the fields, so I don't think we want to create the shortcut.

ash211 replied:
how about a named constructor then, something like `KubernetesDriverSpec.createEmptyWithSparkConf()` or similar?

mccheah replied:
Works for me

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -16,256 +16,130 @@
+    val resolvedDriverJavaOpts = currentDriverSpec
+      .driverSparkConf
+      // We don't need this anymore since we just set the JVM options on the environment
+      .remove(org.apache.spark.internal.config.DRIVER_JAVA_OPTIONS)
```
this filtering should be in some kind of "finalization" step -- want almost all logic in steps

mccheah replied:
The problem with filtering the JVM options in a step is that it's not obvious just from the context of the step itself why the step is doing that. Putting the filtering right next to the application of the JVM options to the container flows better.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -16,256 +16,130 @@
-        resolvedDriverPodBuilder
-      )}.getOrElse(resolvedDriverPodBuilder.build())
+      .build()
     Utils.tryWithResource(
```
this part at the end that actually creates the resolved main container and pod in k8s should be part of some kind of DriverSpecDeployer.  That class would take the given DriverSpec and actually launch it into a k8s cluster.  That should further shorten this method and pull logic out to a more testable place

mccheah replied:
I'm actually not sure how much this buys us in the end. Extracting the logic to a separate module makes `Client#run()` trivial. It would look something like this:

```
  def run(): Unit = {
    var currentDriverSpec = KubernetesDriverSpec.initialSpec(submissionSparkConf)
    // submissionSteps contain steps necessary to take, to resolve varying
    // client arguments that are passed in, created by orchestrator
    for (nextStep <- submissionSteps) {
      currentDriverSpec = nextStep.configureDriver(currentDriverSpec)
    }
    driverRunner.runDriver(currentDriverSpec)
  }
```

The problem with this kind of abstraction is that it just takes the bulk of complexity from one class and moves it to another. We don't get much extra testability out of this as a result because we're just moving the test logic from one test to another. This feels at least to me like "abstraction for the sake of abstraction". What do you think? cc @ifilonenko also for thoughts.

mccheah replied:
Perhaps a better way to think about this is to ask if `ClientSuite` is currently difficult to read. @ash211 @ifilonenko how does that test look right now? If it's testing too many things and it's difficult to read then it's worth creating a separation of concerns. If the test seems like it's testing a reasonably small set of functionality then we should keep the current structure.

(edit: tagged the wrong username)

ifilonenko replied:
@ash211 @mccheah I don't see overly-complex logic in the Client class that would warrant moving the DriverSpec logic into a separate step. I understand your point Andrew, but it might not be wise to overdo the step based architecture at the cost of, as Matt said, just to move complexity from one class to another. However if, as Andrew is saying, we should include a spec building step and separate step to deploy that spec, then we should hold to that paradigm.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -283,100 +157,54 @@ private[spark] class Client(
+    */
+  def main(args: Array[String]): Unit = {
+    val parsedArguments = ClientArguments.fromCommandLineArgs(args)
+    val sparkConf = new SparkConf()
```
comment that this reads from system properties?

ash211 replied:
and reorder to match order of `run` method params

mccheah replied:
The SparkConf constructor's Scaladoc states this sufficiently enough.

mccheah replied:
The fact that it loads from System Properties I mean.

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/Client.scala:
```diff
@@ -283,100 +157,54 @@ private[spark] class Client(
+   /**
+    * Entry point from SparkSubmit in spark-core
+    *
+    *
```
nit: extra newline

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/KubernetesFileUtils.scala:
```diff
@@ -48,4 +50,47 @@ private[spark] object KubernetesFileUtils {
+
+  /**
+   * For the collection of uris, resolves any files as follows:
+   * - Files with scheme file:// are resolved to the download path
```
`the given download path`

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/KubernetesSubmissionStepsOrchestrator.scala:
```diff
@@ -0,0 +1,136 @@
+    submissionSparkConf: SparkConf) {
+
+  // The resource name prefix is derived from the application name, making it easy to connect the
+  // names of the Kubernetes resources from e.g. Kubectl or the Kubernetes dashboard to the
```
lowercase `kubectl`

ash211 commented on resource-managers/kubernetes/core/src/main/scala/org/apache/spark/deploy/kubernetes/submit/submitsteps/KubernetesSubmissionStep.scala:
```diff
@@ -0,0 +1,28 @@
+/**
+ * Represents a step in preparing the Kubernetes driver.
+ */
+private[spark] trait KubernetesSubmissionStep {
```
Just `DriverStep` ? goes with the `InitContainerStep`

mccheah replied:
`DriverStep` sounds pretty vague - we want to indicate that we're configuring the driver somehow. Maybe `DriverConfigurationStep` and `InitContainerConfigurationStep`.

ash211 replied:
Oh I like `DriverConfigurationStep` and `InitContainerConfigurationStep`

ash211 reviewed (commented):
Wow this is so much more readable than before!  I think the concept of steps and the orchestrator assembling steps has worked out really well -- this is much easier to follow and understand than before. 
 Plus the tests seem much more straightforward than before great work!

Besides the little nits, I think adding a `DriverSpecDeployer` or something similar would be the biggest improvement from here.  It would further extract logic to a testable place, and is nice to understand as the followup to creating a spec: should be to deploy the spec.

Again, love this change!

mccheah committed:
Address comments.

mccheah committed:
Merge branch 'submission-steps-refactor' of github.com:apache-spark-on-k8s/spark into submission-steps-refactor

mccheah committed:
Rename class for consistency.

mccheah committed:
Attempt to make spacing consistent.

Multi-line methods should have four-space indentation for arguments that
aren't on the same line as the method call itself... but this is
difficult to do consistently given how IDEs handle Scala multi-line indentation
in most cases.
